### PR TITLE
Host.Context pooling

### PR DESCRIPTION
**Pull request by benaadams**
Host context pooling part from https://github.com/aspnet/AspNetCore/pull/10319

/cc @davidfowl

### CODE REVIEW

File: src/Hosting/Server.Abstractions/src/Internal/IHostContextContainer.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Hosting.Server.Abstractions.Internal
```
Comment (davidfowl) [inline src/Hosting/Server.Abstractions/src/Internal/IHostContextContainer.cs]: pubternal is gone, change the namespace and move this type to public.

Comment (benaadams) [reply src/Hosting/Server.Abstractions/src/Internal/IHostContextContainer.cs]: Moved

**Comment by benaadams** [comment]
Improved comments

**Comment by benaadams** [comment]
Rebased

**Comment by benaadams** [comment]
Rebased to resolve conflict

**Comment by benaadams** [comment]
Superseded by #12385

**Pull request closed without merging**